pallets/allocations/src/lib.rs

```diff
@@ -77,84 +82,99 @@ pub mod pallet {
+		#[pallet::weight(<T as pallet::Config>::WeightInfo::batch(batch.len().try_into().unwrap_or(T::MaxAllocs::get())))]
 		#[transactional]
-		pub fn allocate(
+		pub fn batch(
 			origin: OriginFor<T>,
-			to: T::AccountId,
-			amount: BalanceOf<T>,
-			proof: Vec<u8>,
+			batch: BoundedVec<(T::AccountId, BalanceOf<T>), T::MaxAllocs>,
 		) -> DispatchResultWithPostInfo {
 			Self::ensure_oracle(origin)?;
-			ensure!(
-				!pallet_emergency_shutdown::Pallet::<T>::shutdown(),
-				Error::<T>::UnderShutdown
-			);
-			ensure!(
-				amount >= T::ExistentialDeposit::get().saturating_mul(2u32.into()),
-				Error::<T>::DoesNotSatisfyExistentialDeposit,
-			);
 
-			if amount == Zero::zero() {
-				return Ok(Pays::No.into());
+			// sanity checks
+			let mut full_issuance: BalanceOf<T> = Zero::zero();
+			for (_account, amount) in batch.iter().cloned() {
+				ensure!(
+					amount >= T::ExistentialDeposit::get().saturating_mul(2u32.into()),
+					Error::<T>::DoesNotSatisfyExistentialDeposit,
+				);
+
+				// overflow, so too many coins to allocate
+				full_issuance = full_issuance
+					.checked_add(&amount)
+					.ok_or(Error::<T>::TooManyCoinsToAllocate)?;
 			}
 
-			let coins_already_allocated = Self::coins_consumed();
-			let coins_that_will_be_consumed = coins_already_allocated
-				.checked_add(&amount)
-				.ok_or("Overflow computing coins consumed")?;
+			if full_issuance == Zero::zero() {
+				return Ok(Pays::No.into());
+			}
 
+			let current_supply = T::Currency::total_issuance();
 			ensure!(
-				coins_that_will_be_consumed <= T::MaximumCoinsEverAllocated::get(),
+				current_supply.saturating_add(full_issuance) <= T::MaximumSupply::get(),
 				Error::<T>::TooManyCoinsToAllocate
 			);
 
-			// When using a Perbill type as T::ProtocolFee::get() returns the default way to go is to used the
-			// standard mathematic operands. The risk of {over, under}flow is void as this operation will
-			// effectively take a part of `amount` and thus always produce a lower number. (We use Perbill to
-			// represent percentages)
-			let amount_for_protocol = T::ProtocolFee::get() * amount;
-			let amount_for_grantee = amount.saturating_sub(amount_for_protocol);
-
-			<CoinsConsumed<T>>::put(coins_that_will_be_consumed);
+			// allocate the coins to the proxy account
+			T::Currency::resolve_creating(&T::PalletId::get().into_account(), T::Currency::issue(full_issuance));
+
+			// send to accounts, unfortunately we need to loop again
+			let mut full_protocol: BalanceOf<T> = Zero::zero();
+			for (account, amount) in batch.iter().cloned() {
+				let amount_for_protocol = T::ProtocolFee::get() * amount;
+				let amount_for_grantee = amount.saturating_sub(amount_for_protocol);
+				T::Currency::transfer(
+					&T::PalletId::get().into_account(),
+					&account,
+					amount_for_grantee,
+					ExistenceRequirement::KeepAlive,
+				)?;
+				full_protocol = full_protocol.saturating_add(amount_for_protocol);
```

Comment:
Nice accumulation here. I bet it has caused at least 2x efficiency.

Reply:
yup, most definitely